### ISSUE TITLE
AT_STALE_THREAD_WRITE_OF_PRIMITIVE false positive

### ISSUE THREAD

**Issue by ben-manes** [spotbugs/spotbugs#3310]
```console
Shared primitive variable "written" in one thread may not yield the value of the most recent write from another thread

Bug type AT_STALE_THREAD_WRITE_OF_PRIMITIVE
In class com.github.benmanes.caffeine.jcache.CacheProxy
In method com.github.benmanes.caffeine.jcache.CacheProxy.lambda$putNoCopyOrAwait$8(Object, boolean, Object, CacheProxy$PutResult, Object, Expirable)
Field com.github.benmanes.caffeine.jcache.CacheProxy$PutResult.written
At CacheProxy.java:[line 380]
Another occurrence at CacheProxy.java:[line 394]
```

This case is unfortunately a [bit complex](https://github.com/ben-manes/caffeine/blob/2befdb17e793848da0fba95683d4a041ada5c5e0/jcache/src/main/java/com/github/benmanes/caffeine/jcache/CacheProxy.java#L380), but there is only a single thread able to read and write to this value. The caller is using a capturing lambda in a Map's `compute` to complete with more information about what happened when inside this atomic operation. This way while the map stores a value, additional facts during the operation are communicated to the outside scope.

```java
var result = ...
map.compute(key, (k, oldValue) -> {
    ...
    if (published) {
      result.written = true
    }
    return newValue;
});
return result;
```

**Comment by JuditKnoll** [comment]
Can you please provide a [minimal complete example](https://stackoverflow.com/help/minimal-reproducible-example)?

**Comment by ben-manes** [comment]
Yes, I will try to reduce this and sorry for not doing so originally. In the meantime, you can reproduce this on the Caffeine project by removing the suppression in `gradle/config/spotbugs/exclude.xml` and executing

```console
./gradlew :jcache:spotbugsMain -Dspotbugs --rerun
```

**Comment by gtoison** [comment]
@JuditKnoll I think that the test in https://github.com/spotbugs/spotbugs/pull/3313 reproduces the problem.
I did not investigate it so far, so I'm not sure what might be the problem
